WordPress/gutenberg
# docgen: Optimize README update script

Previously: #15679, #15200

This pull request seeks to optimize and refactor the `bin/update-readmes.js` script. The resulting changes should decrease runtime of this script by **90% or more** for typical usage (baseline of 6.65s to 0.6s `lint-staged` single package or 1.275s complete run average).

The goal here was largely to reduce the delay in `pre-commit` tasks.

**Implementation Notes:**

Influenced by previous optimizations to the Gutenberg build script in #15230, the approach here uses a [`fast-glob`](https://www.npmjs.com/package/fast-glob) stream, allowing compilation to begin even before all files are known.

Like in #15200, the process is now again asynchronous, but incorporates necessary revisions from #15679 to ensure that tokens within a single file can only be replaced in sequence, not in parallel, to avoid conflict. 

It avoids a hard-coded list in favor of reading in the contents of discovered README files to determine whether replacement tokens exist. A "new" syntax was incorporated for files which contain multiple tokens that source from separate files (see "Autogenerated actions", "Autogenerated selectors"). This is really more a convention considered by `update-readmes.js`, and doesn't require any revisions to the `docgen` tool.

Finally, since `lint-staged` will already [provide the staged files as arguments to the script](https://github.com/okonet/lint-staged#why), we can use this to filter packages to those where modifications have occurred.

**Testing Instructions:**

Verify that changes to a file covered by `docgen` will update the corresponding README, including those with multiple or custom "Autogenerated" tokens (e.g. `core-data`).

Verify that the behavior of `lint-staged` is not regressed. You can do this by editing a JavaScript file, staging it (`git add`), then running `npx lint-staged`.
